jupyterhub/app.py

```diff
@@ -1877,16 +1878,23 @@ async def init_roles(self):
                                     "Username %r is not in Authenticator.allowed_users"
                                     % bname
                                 )
-                        roles.add_obj(
-                            db, objname=bname, kind=bearer, rolename=predef_role['name']
+                        Class = orm.get_class(bearer)
+                        orm_obj = Class.find(db, bname)
+                        roles.grant_role(
+                            db, entity=orm_obj, rolename=predef_role['name']
                         )
 
-        # make sure all users, services and tokens have at least one role (update with default)
-        for bearer in role_bearers:
-            Class = orm.get_class(bearer)
-            for obj in db.query(Class):
-                if len(obj.roles) < 1:
-                    roles.update_roles(db, obj=obj, kind=bearer)
+        # make sure that on no admin situation, all roles are reset
+        admin_role = orm.Role.find(db, name='admin')
+        if not admin_role.users:
+            app_log.info(
+                "No admin users found; assuming hub upgrade. Initializing default roles for all entities"
+            )
+            for bearer in role_bearers:
+                Class = orm.get_class(bearer)
+                for obj in db.query(Class):
+                    # if len(obj.roles) < 1: # todo: Should I check if some roles are already assigned?
```

Comment:
I don't think that anyone will be adding roles to the config file yet on upgrade. We can also mention that in docs as well to make sure. Unless we want to allow for new role creation and assignment during upgrade already which I don't think is a good idea. @0mar what do you think?

Reply:
I agree, thanks!